Add infra TorchFunctionModeKey enum and extra TLS state for disabling

soulitzer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #121445
* __->__ #121373

This PR:
- Introduces TorchFunctionModeKey enum (today, only has a single key PROXY) as the torch function counterpart to TorchDispatchModeKey.  They are separate since they are distinct modes.
   - Context: BOTH subclasses and modes set the `_mode_key` attribute with keys in the enum in order to specify to the system that they (1) that they are "infra" and (2) which infra slot in particular they belong to. For modes, this is used to specify that the mode has lower dispatcher preference. For subclasses on the other hand, ordering is actually dictated by the infra mode (and infra subclass?) seeing user subclass and returning NotImplemented, and so this PR shouldn't really need to deal with subclass ordering)

- Introduces an extra state to the disable torch function TLS - Today there are 3 states. In order of decreasing strictness, they are: 1) all enabled, 2) all subclasses disabled, 3) modes and subclass disabled. This PR introduces new state 1.5) only user/non-infra subclasses disabled.
- The behavior of existing APIs do not change. Adds (1) a new context manager to disable torch function subclass without disabling infra subclasses and (2) function to query for the new disabled state (used for testing). After this PR, the API should look like:
   - TorchFunction
     - `torch_function_enabled` (i.e.. `isEnabledTorchFunction`) checks whether BOTH modes and subclasses are enabled
     -  [NEW] `torch_function_infra_subclasses_enabled` (i.e. `isEnabledTorchFunctionInfraSubclasses`)
     - `DisableTorchFunction` disables both modes and subclasses
     - `DisableTorchFunctionSubclass` disables both infra and non-infra subclasses
     - [NEW] `DisableTorchFunctionNonInfraSubclass` disables only non-infra subclasses.
     - `EnableTorchFunction` (deprecated?) enables both modes and subclasses
   - TorchDispatch
      - `DisableTorchDispatch` (i.e. `no_dispatch`) disables both modes and subclasses by excluding the python keys 
      - `dispatch_mode_enabled` any modes on the stack (including infra modes)

See https://github.com/pytorch/pytorch/issues/120124#issuecomment-1981346560 for context
   
cc @bdhirsh

soulitzer committed:
Add infra TorchFunctionModeKey enum and extra TLS state for disabling

[ghstack-poisoned]

soulitzer committed:
Update on "Add infra TorchFunctionModeKey enum and extra TLS state for disabling"




[ghstack-poisoned]

soulitzer committed:
Update on "Add infra TorchFunctionModeKey enum and extra TLS state for disabling"


This PR:
- Introduces TorchFunctionModeKey enum (today, only has a single key PROXY) as the torch function counterpart to TorchDispatchModeKey.  They are separate since they are distinct modes.
   - Context: BOTH subclasses and modes set the `_mode_key` attribute with keys in the enum in order to specify to the system that they (1) that they are "infra" and (2) which infra slot in particular they belong to. For modes, this is used to specify that the mode has lower dispatcher preference. For subclasses on the other hand, ordering is actually dictated by the infra mode (and infra subclass?) seeing user subclass and returning NotImplemented, and so this PR shouldn't really need to deal with subclass ordering)

- Introduces extra state to the disable torch function TLS - Today there are 3 states. In order of decreasing strictness, they are: 1) all enabled, 2) all subclasses disabled, 3) modes and subclass disabled. This PR introduces new state 1.5) only user/non-infra subclasses disabled.
- The behavior of existing PRs do not change. Adds (1) a new context manager to disable torch function subclass without disabling infra subclasses and (2) function to query for the new disabled state. After this PR, the API should look like:
   - TorchFunction
     - `torch_function_enabled` (i.e.. `isEnabledTorchFunction`) checks whether BOTH modes and subclasses are enabled
     -  [NEW] `torch_function_non_infra_subclasses_enabled` (i.e. `isEnabledTorchFunctionNonInfraSubclasses`)
     - `DisableTorchFunction` disables both modes and subclasses
     - `DisableTorchFunctionSubclass` disables only subclasses
     - [NEW] `DisableTorchFunctionNonInfraSubclass` disables only non-infra subclasses.
     - `EnableTorchFunction` (deprecated?) enables both modes and subclasses
   - TorchDispatch
      - `DisableTorchDispatch` (i.e. `no_dispatch`) disables both modes and subclasses by excluding the python keys 
      - `dispatch_mode_enabled` any modes on the stack (including infra modes)
   


[ghstack-poisoned]

soulitzer committed:
Update on "Add infra TorchFunctionModeKey enum and extra TLS state for disabling"


This PR:
- Introduces TorchFunctionModeKey enum (today, only has a single key PROXY) as the torch function counterpart to TorchDispatchModeKey.  They are separate since they are distinct modes.
   - Context: BOTH subclasses and modes set the `_mode_key` attribute with keys in the enum in order to specify to the system that they (1) that they are "infra" and (2) which infra slot in particular they belong to. For modes, this is used to specify that the mode has lower dispatcher preference. For subclasses on the other hand, ordering is actually dictated by the infra mode (and infra subclass?) seeing user subclass and returning NotImplemented, and so this PR shouldn't really need to deal with subclass ordering)

- Introduces extra state to the disable torch function TLS - Today there are 3 states. In order of decreasing strictness, they are: 1) all enabled, 2) all subclasses disabled, 3) modes and subclass disabled. This PR introduces new state 1.5) only user/non-infra subclasses disabled.
- The behavior of existing PRs do not change. Adds (1) a new context manager to disable torch function subclass without disabling infra subclasses and (2) function to query for the new disabled state. After this PR, the API should look like:
   - TorchFunction
     - `torch_function_enabled` (i.e.. `isEnabledTorchFunction`) checks whether BOTH modes and subclasses are enabled
     -  [NEW] `torch_function_non_infra_subclasses_enabled` (i.e. `isEnabledTorchFunctionNonInfraSubclasses`)
     - `DisableTorchFunction` disables both modes and subclasses
     - `DisableTorchFunctionSubclass` disables only subclasses
     - [NEW] `DisableTorchFunctionNonInfraSubclass` disables only non-infra subclasses.
     - `EnableTorchFunction` (deprecated?) enables both modes and subclasses
   - TorchDispatch
      - `DisableTorchDispatch` (i.e. `no_dispatch`) disables both modes and subclasses by excluding the python keys 
      - `dispatch_mode_enabled` any modes on the stack (including infra modes)
   


[ghstack-poisoned]

soulitzer committed:
Update on "Add infra TorchFunctionModeKey enum and extra TLS state for disabling"


This PR:
- Introduces TorchFunctionModeKey enum (today, only has a single key PROXY) as the torch function counterpart to TorchDispatchModeKey.  They are separate since they are distinct modes.
   - Context: BOTH subclasses and modes set the `_mode_key` attribute with keys in the enum in order to specify to the system that they (1) that they are "infra" and (2) which infra slot in particular they belong to. For modes, this is used to specify that the mode has lower dispatcher preference. For subclasses on the other hand, ordering is actually dictated by the infra mode (and infra subclass?) seeing user subclass and returning NotImplemented, and so this PR shouldn't really need to deal with subclass ordering)

- Introduces an extra state to the disable torch function TLS - Today there are 3 states. In order of decreasing strictness, they are: 1) all enabled, 2) all subclasses disabled, 3) modes and subclass disabled. This PR introduces new state 1.5) only user/non-infra subclasses disabled.
- The behavior of existing APIs do not change. Adds (1) a new context manager to disable torch function subclass without disabling infra subclasses and (2) function to query for the new disabled state (used for testing). After this PR, the API should look like:
   - TorchFunction
     - `torch_function_enabled` (i.e.. `isEnabledTorchFunction`) checks whether BOTH modes and subclasses are enabled
     -  [NEW] `torch_function_infra_subclasses_enabled` (i.e. `isEnabledTorchFunctionInfraSubclasses`)
     - `DisableTorchFunction` disables both modes and subclasses
     - `DisableTorchFunctionSubclass` disables both infra and non-infra subclasses
     - [NEW] `DisableTorchFunctionNonInfraSubclass` disables only non-infra subclasses.
     - `EnableTorchFunction` (deprecated?) enables both modes and subclasses
   - TorchDispatch
      - `DisableTorchDispatch` (i.e. `no_dispatch`) disables both modes and subclasses by excluding the python keys 
      - `dispatch_mode_enabled` any modes on the stack (including infra modes)

See https://github.com/pytorch/pytorch/issues/120124#issuecomment-1981346560 for context
   
cc bdhirsh 


[ghstack-poisoned]

soulitzer committed:
Update on "Add infra TorchFunctionModeKey enum and extra TLS state for disabling"


This PR:
- Introduces TorchFunctionModeKey enum (today, only has a single key PROXY) as the torch function counterpart to TorchDispatchModeKey.  They are separate since they are distinct modes.
   - Context: BOTH subclasses and modes set the `_mode_key` attribute with keys in the enum in order to specify to the system that they (1) that they are "infra" and (2) which infra slot in particular they belong to. For modes, this is used to specify that the mode has lower dispatcher preference. For subclasses on the other hand, ordering is actually dictated by the infra mode (and infra subclass?) seeing user subclass and returning NotImplemented, and so this PR shouldn't really need to deal with subclass ordering)

- Introduces an extra state to the disable torch function TLS - Today there are 3 states. In order of decreasing strictness, they are: 1) all enabled, 2) all subclasses disabled, 3) modes and subclass disabled. This PR introduces new state 1.5) only user/non-infra subclasses disabled.
- The behavior of existing APIs do not change. Adds (1) a new context manager to disable torch function subclass without disabling infra subclasses and (2) function to query for the new disabled state (used for testing). After this PR, the API should look like:
   - TorchFunction
     - `torch_function_enabled` (i.e.. `isEnabledTorchFunction`) checks whether BOTH modes and subclasses are enabled
     -  [NEW] `torch_function_infra_subclasses_enabled` (i.e. `isEnabledTorchFunctionInfraSubclasses`)
     - `DisableTorchFunction` disables both modes and subclasses
     - `DisableTorchFunctionSubclass` disables both infra and non-infra subclasses
     - [NEW] `DisableTorchFunctionNonInfraSubclass` disables only non-infra subclasses.
     - `EnableTorchFunction` (deprecated?) enables both modes and subclasses
   - TorchDispatch
      - `DisableTorchDispatch` (i.e. `no_dispatch`) disables both modes and subclasses by excluding the python keys 
      - `dispatch_mode_enabled` any modes on the stack (including infra modes)

See https://github.com/pytorch/pytorch/issues/120124#issuecomment-1981346560 for context
   
cc bdhirsh 


[ghstack-poisoned]

jbschlosser reviewed (commented):
Thanks for the PR! The design looks like what I expected so I'm good with it. will defer to the experts in this area for the final sign-off though

soulitzer committed:
Update on "Add infra TorchFunctionModeKey enum and extra TLS state for disabling"


This PR:
- Introduces TorchFunctionModeKey enum (today, only has a single key PROXY) as the torch function counterpart to TorchDispatchModeKey.  They are separate since they are distinct modes.
   - Context: BOTH subclasses and modes set the `_mode_key` attribute with keys in the enum in order to specify to the system that they (1) that they are "infra" and (2) which infra slot in particular they belong to. For modes, this is used to specify that the mode has lower dispatcher preference. For subclasses on the other hand, ordering is actually dictated by the infra mode (and infra subclass?) seeing user subclass and returning NotImplemented, and so this PR shouldn't really need to deal with subclass ordering)

- Introduces an extra state to the disable torch function TLS - Today there are 3 states. In order of decreasing strictness, they are: 1) all enabled, 2) all subclasses disabled, 3) modes and subclass disabled. This PR introduces new state 1.5) only user/non-infra subclasses disabled.
- The behavior of existing APIs do not change. Adds (1) a new context manager to disable torch function subclass without disabling infra subclasses and (2) function to query for the new disabled state (used for testing). After this PR, the API should look like:
   - TorchFunction
     - `torch_function_enabled` (i.e.. `isEnabledTorchFunction`) checks whether BOTH modes and subclasses are enabled
     -  [NEW] `torch_function_infra_subclasses_enabled` (i.e. `isEnabledTorchFunctionInfraSubclasses`)
     - `DisableTorchFunction` disables both modes and subclasses
     - `DisableTorchFunctionSubclass` disables both infra and non-infra subclasses
     - [NEW] `DisableTorchFunctionNonInfraSubclass` disables only non-infra subclasses.
     - `EnableTorchFunction` (deprecated?) enables both modes and subclasses
   - TorchDispatch
      - `DisableTorchDispatch` (i.e. `no_dispatch`) disables both modes and subclasses by excluding the python keys 
      - `dispatch_mode_enabled` any modes on the stack (including infra modes)

See https://github.com/pytorch/pytorch/issues/120124#issuecomment-1981346560 for context
   
cc bdhirsh 


[ghstack-poisoned]

bdhirsh commented on aten/src/ATen/PythonTorchFunctionTLS.h:
```diff
@@ -5,7 +5,14 @@
-enum TorchFunctionDisabledState { ENABLED, SUBCLASSES_DISABLED, ALL_DISABLED };
+enum class TorchFunctionModeKey : int8_t { PROXY, NUM_MODE_KEYS };
+
+enum TorchFunctionDisabledState {
```
maybe add a comment calling out that order matters?

bdhirsh commented on test/test_overrides.py:
```diff
@@ -1506,6 +1504,51 @@ class A(torch.Tensor):
 
+        counter = [0]
+
+        class Foo:
```
Are these changes enough for you to add an E2E test of a subclass that has both a torch_function and torch_dispatch impl, that we can run torch.compile on? (that might also be a useful test to add in this PR)

soulitzer replied:
Ah do you mean to make sure that we can now use this disable-non-infra torch function subclass context manager to ensure that torch function is disabled? This is currently tested in the next PR, but I guess I can move it to this one.

bdhirsh replied:
np if you have a similar test further in the stack

bdhirsh commented:
The changes look good to me. Would be good to get a review from @zou3519 too.

soulitzer committed:
Update on "Add infra TorchFunctionModeKey enum and extra TLS state for disabling"


This PR:
- Introduces TorchFunctionModeKey enum (today, only has a single key PROXY) as the torch function counterpart to TorchDispatchModeKey.  They are separate since they are distinct modes.
   - Context: BOTH subclasses and modes set the `_mode_key` attribute with keys in the enum in order to specify to the system that they (1) that they are "infra" and (2) which infra slot in particular they belong to. For modes, this is used to specify that the mode has lower dispatcher preference. For subclasses on the other hand, ordering is actually dictated by the infra mode (and infra subclass?) seeing user subclass and returning NotImplemented, and so this PR shouldn't really need to deal with subclass ordering)

- Introduces an extra state to the disable torch function TLS - Today there are 3 states. In order of decreasing strictness, they are: 1) all enabled, 2) all subclasses disabled, 3) modes and subclass disabled. This PR introduces new state 1.5) only user/non-infra subclasses disabled.
- The behavior of existing APIs do not change. Adds (1) a new context manager to disable torch function subclass without disabling infra subclasses and (2) function to query for the new disabled state (used for testing). After this PR, the API should look like:
   - TorchFunction
     - `torch_function_enabled` (i.e.. `isEnabledTorchFunction`) checks whether BOTH modes and subclasses are enabled
     -  [NEW] `torch_function_infra_subclasses_enabled` (i.e. `isEnabledTorchFunctionInfraSubclasses`)
     - `DisableTorchFunction` disables both modes and subclasses
     - `DisableTorchFunctionSubclass` disables both infra and non-infra subclasses
     - [NEW] `DisableTorchFunctionNonInfraSubclass` disables only non-infra subclasses.
     - `EnableTorchFunction` (deprecated?) enables both modes and subclasses
   - TorchDispatch
      - `DisableTorchDispatch` (i.e. `no_dispatch`) disables both modes and subclasses by excluding the python keys 
      - `dispatch_mode_enabled` any modes on the stack (including infra modes)

See https://github.com/pytorch/pytorch/issues/120124#issuecomment-1981346560 for context
   
cc bdhirsh 


[ghstack-poisoned]

soulitzer committed:
Update on "Add infra TorchFunctionModeKey enum and extra TLS state for disabling"


This PR:
- Introduces TorchFunctionModeKey enum (today, only has a single key PROXY) as the torch function counterpart to TorchDispatchModeKey.  They are separate since they are distinct modes.
   - Context: BOTH subclasses and modes set the `_mode_key` attribute with keys in the enum in order to specify to the system that they (1) that they are "infra" and (2) which infra slot in particular they belong to. For modes, this is used to specify that the mode has lower dispatcher preference. For subclasses on the other hand, ordering is actually dictated by the infra mode (and infra subclass?) seeing user subclass and returning NotImplemented, and so this PR shouldn't really need to deal with subclass ordering)

- Introduces an extra state to the disable torch function TLS - Today there are 3 states. In order of decreasing strictness, they are: 1) all enabled, 2) all subclasses disabled, 3) modes and subclass disabled. This PR introduces new state 1.5) only user/non-infra subclasses disabled.
- The behavior of existing APIs do not change. Adds (1) a new context manager to disable torch function subclass without disabling infra subclasses and (2) function to query for the new disabled state (used for testing). After this PR, the API should look like:
   - TorchFunction
     - `torch_function_enabled` (i.e.. `isEnabledTorchFunction`) checks whether BOTH modes and subclasses are enabled
     -  [NEW] `torch_function_infra_subclasses_enabled` (i.e. `isEnabledTorchFunctionInfraSubclasses`)
     - `DisableTorchFunction` disables both modes and subclasses
     - `DisableTorchFunctionSubclass` disables both infra and non-infra subclasses
     - [NEW] `DisableTorchFunctionNonInfraSubclass` disables only non-infra subclasses.
     - `EnableTorchFunction` (deprecated?) enables both modes and subclasses
   - TorchDispatch
      - `DisableTorchDispatch` (i.e. `no_dispatch`) disables both modes and subclasses by excluding the python keys 
      - `dispatch_mode_enabled` any modes on the stack (including infra modes)

See https://github.com/pytorch/pytorch/issues/120124#issuecomment-1981346560 for context
   
cc bdhirsh 


[ghstack-poisoned]

soulitzer commented:
Discussed this some more with @bdhirsh offline: we're probably fine with not landing this for now to avoid introducing the concept of infra TorchFunctionMode/subclass, and just saying that if you are dealing with Proxy in aot autograd, you should not rely on its `__torch_function__` to do the call function. This tradeoff seems to make sense because Proxy logic hasn't really changed all that much in the past year, and its really only depended on in a single place AFAIK, which already been patched. If we realize that maintaining this invariant is too laborious, we can always go back to this PR.
